opennebula/shared_schemas.go

```diff
@@ -132,6 +132,14 @@ func diskSchema(customFields ...map[string]*schema.Schema) *schema.Schema {
 	}
 }
 
+func featuresSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:        schema.TypeMap,
```

Comment:
Not sure for `TypeMap`. Unless we want to pass some custom elements, the doc already describe us all the possibilities.
TypeSet may be a better choice ?
For instance, take a look at virtual network AR to see how TypeSet are used.

Reply:
Managed to get it working with `TypeSet` (not committed jet).
But I don't know how to do the `Read` and `Update` code.
The CONTEXT section would be something similar I guess but unfortunately it looks like it's also missing there.

Reply:
I didn't test the code below, it's just a really big picture that I quickly wrote so you'll probably have to fix problems, refactor parts.

To read you'll need to work with datas retrieved from ONE to give them to Terraform.
So, you have to retrieve the `FEATURES` vector in the `TEMPLATE` section of the template resource, something like this:
```
			features, err := template.Template.Get("FEATURES")
			if err != nil {
				return err
			}
```
From here, you'll need to populate a `map[string]interface{}`, so for each attributes of the `FEATURES` section, you'll have to retrieve the key and the value, put them in the map:
```

	value1, _ := disk.Get("KEY1")
	value2, _ := disk.Get("KEY2")
        ...

	features := map[string]interface{}{
		"schema_key_1":        value1,
		"schema_key_2":        value2,
                ...
	}

```
Then you'll need to call set the map key:
```
   err := d.Set("features", []map[string]interface{features})
``` 
Fill free to refactor to make something the more simple and readable as you can.

For the update, you'll need to add some code inside at test that check if values is schemas values has changed in the Update function.
In this section you'll need to build the `FEATURES` section, add it to a template, and then send the template to OpenNebula via the update method.
In fact there is some existing code you'll need to integrate with, look around `tags` section in the update method, this will be something like:
```
	if d.HasChange("features") {
		newTpl.Del("FEATURES")

		features := d.Get("features").(*schema.Set).List()
		for i := 0; i < len(features); i++ {
		        featuresMap := features[i].(map[string]interface{})

                        // maybe that you can reuse some code already wrote in the create method instead of adapting the code below
                        value1 := featuresMap["schema_key_1"].(int) // cast to the type to match the type described in the schema
                        value2 := featuresMap["schema_key_2"].(string)
                        
			featuresVec := newTpl.AddVector("FEATURES")
			
			featuresVec.AddPair("schema_key_1", value1)
			featuresVec.AddPair("schema_key_2", value2)
		}

		update = true
	}
```

Reply:
Thanks for your support!

`Update` seems to work now.
Not sure about the `Read` stuff. I hope it's also fine to do it in `shared_schemas`. To be honest I don't exactly know how to test/verify this on my own.

The 2 `FEATURES` `iothread` and `hyperv` seems to be not implemented in vendor/github.com/OpenNebula/one/src/oca/go/src/goca/schemas/vm/keys/template.go. That's why I skipped them.

Please review and give feedback again.